packages/store-consumer/src/experimental/WithWorld.sol

```diff
@@ -6,20 +6,31 @@ import { ResourceId } from "@latticexyz/store/src/ResourceId.sol";
 import { ResourceAccess } from "@latticexyz/world/src/codegen/tables/ResourceAccess.sol";
 import { IBaseWorld } from "@latticexyz/world/src/codegen/interfaces/IBaseWorld.sol";
 import { WorldResourceIdLib } from "@latticexyz/world/src/WorldResourceId.sol";
+import { WorldContextConsumer } from "@latticexyz/world/src/WorldContext.sol";
+import { System } from "@latticexyz/world/src/System.sol";
 
+import { Context } from "./Context.sol";
 import { WithStore } from "./WithStore.sol";
 
-abstract contract WithWorld is WithStore {
+abstract contract WithWorld is WithStore, System {
   bytes14 public immutable namespace;
 
   error WithWorld_RootNamespaceNotAllowed();
   error WithWorld_NamespaceAlreadyExists();
   error WithWorld_NamespaceDoesNotExists();
   error WithWorld_CallerHasNoNamespaceAccess();
+  error WithWorld_CallerIsNotWorld();
+
+  modifier onlyWorld() {
+    if (!_callerIsWorld()) {
+      revert WithWorld_CallerIsNotWorld();
+    }
+    _;
+  }
 
   modifier onlyNamespace() {
-    address sender = _msgSender();
-    if (!ResourceAccess.get(getNamespaceId(), sender)) {
+    // We use WorldContextConsumer directly as we already know the world is the caller
+    if (!_callerIsWorld() || !ResourceAccess.get(getNamespaceId(), WorldContextConsumer._msgSender())) {
```

Comment:
so if my EOA has access to the namespace I still have to call the token system through the world correct?